src/mesh/BoundingBox.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+#include "logging/Logger.hpp"
+#include "mesh/Vertex.hpp"
+#include <vector>
+
+namespace precice {
+namespace mesh {
+
+/// An axis-aligned bounding box around a (partition of a) mesh
+class BoundingBox {
+
+public:
+  /***
+  * 
+  * @brief Constructor.
+  * 
+  * @param[in] bounds Vector of minimum and maximum points in each dimension
+  * 
+  */
+  BoundingBox(std::vector<double> bounds);
+  
+  /***
+  * 
+  * @brief Constructor.
+  * 
+  * @param[in] dimension Dimension of the bounding box
+  * 
+  */
+  BoundingBox(int dimension);
+
+  /// Default constructor
+  BoundingBox();
+
+  /// Copy Constructor
+  BoundingBox(const BoundingBox &bb);
+
+  /// Comparison Operator
+  bool operator==(const BoundingBox& otherBB) const;
+
+  /// Bounding box factory function
+  static BoundingBox createFromData(std::vector<double> bounds);
+
+  /// Expand bounding box using another bounding box
+  void expandTo(const BoundingBox &otherBB);
+
+  /// Expand bounding box using vertices
+  void expandTo(const Vertex& vertices);
+
+  /// Expand bounding box using value
+  void expandTo(double value);
+
+  /// Increase the size of bounding box by safety margin
+  void addSafetyMargin(double safetyFactor);
+
+  /// Checks if vertex in contained in _bb
+  bool isVertexInBB(const Vertex &vertex) const;
+
+  /// Checks whether two bounding boxes are overlapping
+  bool overlapping(const BoundingBox &otherBB);
+
+  /**
+   * @brief Returns the Center Of Gravity of the mesh
+   *
+   * Returns a vector of doubles, size d, each dimension computed as
+   * cog =  (max - min) / 2 + min
+   */
+  std::vector<double> getCOG() const;
+
+  /// Calculate the area of bounding box
+  double getArea(std::vector<bool> deadAxis);
+
+  /// Return data as std::vector
+  const std::vector<double> dataVector() const;
```

Comment:
> currently using only the std::vector version, sending and receiving can be performed.

Receiving won't be possible with `data()` and `size()` due to the const specifier, but sending should work fine.
For receiving, you can create a `std::vector<double>`, receive to it and pass/move that to the BoundingBox constructor.

From an interface perspective, the only thing that is missing is the access to the BoundingBox data which is required for sending. Both `data()` and `dataVector()` provide this access, whereas the former is a minimal approach. `size()` is not even required as it simply returns `2*getDimensions()`.


@uekerman opinions on this?

Reply:
Yes, let's keep it minimal.

> For receiving, you can create a std::vector<double>, receive to it and pass/move that to the BoundingBox constructor.

Sounds good.

And for sending either `data()` or `dataVector`, no strong opinion which one, but not both.

Reply:
Okay, therefore I will switch it to `data()` then. Since it is not used other than the communication, I won't add the `size()` function, as Frederic mentioned, `getDimension()` should work fine.